services/outputhost/replicaconnection.go

```diff
@@ -186,124 +192,183 @@ func (conn *replicaConnection) readMessagesPump() {
-							"msgSeqNum":       msgSeqNum,
-							"expectedSeqNum":  expectedSeqNum,
-							"skippedMessages": skippedMessages,
-						}).Error("sequence number out of order")
-					}
-				} else {
-					// T471157 For timers, do not signal discontinuities to ack manager, since discontinuities are frequent
-					msgSeqNum = 0
+				// compute the 'visibility-time' of the message, taking into account the specified 'delay', if any.
+				visibilityTime := msg.Message.GetEnqueueTimeUtc()
+
+				if doDelay && visibilityTime > 0 {
+					visibilityTime += delayNanos
 				}
 
-				// update the lastSeqNum to this value
-				lastSeqNum = msg.Message.GetSequenceNumber()
+				now := time.Now().UnixNano()
 
-				// convert this to an outMessage
-				cMsg := cherami.NewConsumerMessage()
-				cMsg.EnqueueTimeUtc = msg.Message.EnqueueTimeUtc
-				cMsg.Payload = msg.Message.Payload
+				// check if the messages have already 'expired'; ie, if it falls outside the
+				// skip-older window. ignore (ie, don't skip any messages), if skip-older is '0'.
+				// NB: messages coming from KFC may not have enqueue-time set; the logic below
+				// ignores (ie, does not skip) messages that have no/zero enqueue-time.
+				if doSkipOlder && (visibilityTime > 0) && (visibilityTime < (now - skipOlderNanos)) {
 
-				cMsg.AckId = common.StringPtr(conn.extCache.ackMgr.getNextAckID(storeHostAddress(msg.GetAddress()), msgSeqNum))
-				// write the message to the msgsCh so that it can be delivered
-				// after being stored on the cache.
-				// 1. either there are no listeners
-				// 2. the buffer is full
-				// Wait until the there are some listeners or we are shutting down
-				select {
-				case conn.msgsCh <- cMsg:
-					// written successfully. Now accumulate credits
-					// TODO: one message is now assumed to take one credit
-					// we might need to change it to be based on size later.
-					// we accumulate a bunch of credits and then send it in a batch to store
-					conn.updateSuccessfulSendToMsgsCh(&localReadMsgs, int64(len(cMsg.Payload.GetData())))
-				default:
-					// we were unable to write it above which probably means the channel is full
-					// now do it in a blocking way except shutdown.
+					conn.skipOlderMsgs++
+					continue loop
+				}
+
+				// check if this is a delayed-cg, and if so delay messages appropriately
+				// compute visibility time based on the enqueue-time and specified delay
+				if doDelay && visibilityTime > now {
+
+					delayTimer.Reset(time.Duration(visibilityTime - now))
+
+					// wait unil the message should be made 'visible'
 					select {
-					case conn.msgsCh <- cMsg:
-						// written successfully. Now accumulate credits
-						conn.updateSuccessfulSendToMsgsCh(&localReadMsgs, int64(len(cMsg.Payload.GetData())))
-					// TODO: Make sure we listen on the close channel if and only if, all the
-					// consumers are gone as well. (i.e, separate out the close channel).
+					case <-delayTimer.C: // sleep until delay expiry
```

Comment:
There is a bit of danger here:

1. I suspect that our idle detection logic will kill the connection if it sits here too long. 
2. A firewall admin can give us a very bad day with a small policy change. Long lived zero-throughput connections often don't resume properly. 

I don't think that we should pursue a design change, but you ought to test the idle connection timeout thing once, probably by adjusting the idle timeout to a suitably low value.